### ISSUE TITLE
Cannot use more than 65 variables with TaylorN

### ISSUE THREAD

**Issue by PerezHz** [JuliaDiff/TaylorSeries.jl#85]
Was just doing some work on my machine with TaylorN (OS X 10.10.5, julia-0.5; TaylorSeries v0.4.0), and then I came across some unexpected behaviour:
```julia
julia> using TaylorSeries
julia> δv = set_variables("δ", order=1, numvars=65);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 𝒪(‖x‖²) #this is OK
julia> δv = set_variables("δ", order=1, numvars=66);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₂ + 𝒪(‖x‖²) #the subscript on the δ should be 1 instead of 2
julia> δv = set_variables("δ", order=1, numvars=70);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₆ + 𝒪(‖x‖²) #the subscript on the δ should be 1 instead of 6
julia> δv = set_variables("δ", order=1, numvars=5);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 𝒪(‖x‖²) #this is OK
julia> δv = set_variables("δ", order=1, numvars=65);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 𝒪(‖x‖²) #this is OK
```

There seems to be a problem with `TaylorN` when `order==1` and `numvars>65`...

If I use `order=2` instead of `order=1`, everything works fine (this is the workaround I'm using):
```julia
julia> δv = set_variables("δ", order=2, numvars=65);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 1.0 δ₁² + 𝒪(‖x‖³) #OK
julia> δv = set_variables("δ", order=2, numvars=66);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 1.0 δ₁² + 𝒪(‖x‖³) #OK
julia> δv = set_variables("δ", order=2, numvars=70);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 1.0 δ₁² + 𝒪(‖x‖³) #OK
julia> δv = set_variables("δ", order=2, numvars=5);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 1.0 δ₁² + 𝒪(‖x‖³) #OK
julia> δv = set_variables("δ", order=2, numvars=65);
julia> (1+δv[1])^2
 1.0 + 2.0 δ₁ + 1.0 δ₁² + 𝒪(‖x‖³) #OK
```

After testing in JuliaBox (julia-0.5, TaylorSeries v0.4.0) the results are the same...

**Comment by dpsanders** [comment]
Hmm. Can you try looking at the actual coefficients to see if it's just a problem with the `show` method, or if it's a bug in the actual calculation?

**Comment by PerezHz** [comment]
It seems that the problem has to do with the actual calculation:

```julia
julia> using TaylorSeries

julia> δv = set_variables("δ", order=1, numvars=66);

julia> a = (1+δv[1])^2
 1.0 + 2.0 δ₂ + 𝒪(‖x‖²)

julia> a.coeffs[2].coeffs
66-element Array{Float64,1}:
 0.0
 2.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 ⋮  
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
```

The first element of the last array should be the one with the `2.0`, but instead the second element is the one with `2.0`

**Comment by lbenet** [comment]
> The first element of a.coeffs[2].coeffs should be the one with the 2.0, but instead the second element is the one with 2.0

This seems to be a bug; I'm taking a look what is going on. The weird thing is that it occurs with some number of variables, but with other not. So I would guess it is a subtle issue of memory mapping.

**Comment by lbenet** [comment]
I found the problem!!! It is related to the way we construct the hash tables and specifically `index_table`. Simply stated, the way we hash the multi-indices grows too fast and then the modular arithmetic of `Int64` comes into play. For the case in question, different homogeneous polynomials are hashed identically, and the last one (in the way we number them) is used.

I'm trying to workout a solution...

**Comment by dpsanders** [comment]
Ouch! I guess that's my fault. One possible solution is to use `Int128`, though that will slow everything down a lot.

Maybe it's time to think more seriously about using a sparse representation as in https://github.com/daviddelaat/MultiPoly.jl

**Comment by lbenet** [comment]
The big change was done in b16e006; I was thinking on a simple linear indexing. We actually don't care about *how* the coefficients are hashed, but that it is unique.

**Comment by dpsanders** [comment]
What do you mean by a simple linear indexing? Yes, definitely I agree with you.

**Comment by lbenet** [comment]
I mean something like simply counting the distinct elements of `coeff_table`. I'm already testing this.

**Comment by dpsanders** [comment]
Isn't the problem always the same -- how to do the indexing efficiently?

**Comment by lbenet** [comment]
Yes, you are write. The subtlety I forgot is not only how to define such an index, but go back from the actual coefficients...

**Comment by dpsanders** [comment]
In any case, is it really feasible to do a non-trivial calculation with more than 64 variables??

**Comment by lbenet** [comment]
Well, you see it was... 😄

The current idea I have, which is not working yet, is to hash the vector of coefficients `hash(v)`, and convert (if needed) the resulting `UInt64` to an `Int64`.

**Comment by dpsanders** [comment]
I think we already tried that. In any case, that's basically just the same as having a dictionary and indexing into the dictionary with the vector of coefficients (which is certainly worth trying).

**Comment by dpsanders** [comment]
Well it would only be possible with a sparse representation, which brings us back to the dictionary idea (which I think is what `MultivariatePolynomials.jl` uses -- can we just use that underneath?)

**Comment by lbenet** [comment]
A naive implementation using `hash(v)` doesn't work. The problem is to ensure that `hash(v)+hash(w)` is equal to `hash(v+w)`, with `v` and `w` vectors with the indices (powers) that identify a monomial.

**Comment by dpsanders** [comment]
I think the easiest thing is just to use a dictionary whose keys are the monomial vectors. Is there something wrong with that idea?

**Comment by lbenet** [comment]
I agree with you; yet, that may have penalties (memory) which may hinder performance.

**Comment by dpsanders** [comment]
So let's go back to the currently-implemented idea: map the monomials to an integer. But the *type* of integer may change; it can be a parameter of the type, i.e. in `hash_tables.jl`, make a `HashTables` type something like this:

```
immutable HashTables{IntType}
    coeff_table :: Array{Array{Array{IntType,1},1},1}
    index_table :: Array{Array{IntType,1},1}
    size_table :: Array{IntType,1}
    pos_table :: Array{Dict{IntType,IntType},1}
end 
```
Now you just generate the `HashTables` object with the correct `IntType` that can hold an integer that's large enough: `Int64`, `Int128` or `BigInt`. 

Of course, that will be slow for anything that's not `Int64`, but that's what you get for using so many variables... (Can Mathematica handle lots of variables?)

**Comment by dpsanders** [comment]
See [this StackOverflow question](http://math.stackexchange.com/questions/1130917/how-to-efficiently-represent-and-manipulate-polynomials-in-software) and [this presentation linked there](http://www.maia.ub.es/dsg/sam07/sam07poly.pdf) (which is now in a different place than is linked).

**Comment by dpsanders** [comment]
And [this](http://www.issac-conference.org/2010/assets/SoftwareDemos/PearceFinal.pdf)

**Comment by lbenet** [comment]
I just pushed a PR which addresses this issue. 

The whole trick consists in hashing the vectors of integers as we did before, but using `UInt128` arithmetic. This is certainly too large, I admit, but there is no huge penalty in memory, and it allows @PerezHz to go on.

See #86

**Comment by lbenet** [comment]
A solution is proposed in #86, but as discussed here and there, the solution has an impact on performance. For the time being, I think the best is to look for other alternatives to what is currently done.

I'll leave this open as a remainder.

**Comment by lbenet** [comment]
I note here that, due to changes in #87, I updated #86 so it works with the current organization of files.

**Comment by PerezHz** [comment]
Thank you so much! This is really helpful!

**Comment by PerezHz** [comment]
Should this branch rebase master, now that PRs #90 and #91 have been merged?

**Comment by lbenet** [comment]
Yes, I'll do it right away!

**Comment by dpsanders** [comment]
I think the solution to this is to be able to specify `Int64` or `Int128` (or even `Int32` etc.) as an additional type parameter.

**Comment by patkwee** [comment]
Hi,
I have run into this issue after debugging for 3 hours. Is there a structural solution in sight? As a containment I think it would be helpful if  `set_variables` would throw an error if you are using bad variable number and order combinations that lead to this issue.

**Comment by lbenet** [comment]
Sorry that you stumble onto this known problem and it cost you so much time, and thanks for bringing it back to our minds...

As long as we are using `Int`s (or `UInt`s) to index the hash tables, we will have these problem, simply because they are implemented as modular arithmetic. It is also difficult to catch the problems because, with some bad luck, the combination of variables and order make you end-up with a (wrong) positive integer... Using `BigInt`s avoids the problem completly, but it is too costly in terms of performance, so I think that is no alternative.

Two other options occur to me:
- Using the current scheme for the hash tables, somehow be able to catch when an error occurs; e.g., use [SaferIntegers.jl](https://github.com/JeffreySarnoff/SaferIntegers.jl) to check that the `Int` have now overflown. I have no idea about the impact in performance, though i think it should be small, and only when the tables are built.
- Use the improved hashing tables mentioned in a paper by Berz [[AIP Conference Proceedings 177, 275 (1988)](https://doi.org/10.1063/1.37800)]  by introducing two more tables. This does not solve the problem, but allows to use more variables and higher orders.

The first one, I think, is easier to implement, and is non breaking; the second one seems a more stable and better implementation, but it will involve lots of changes around.

Would you like to help? :-)

**Comment by patkwee** [comment]
Let me see if I correctly understand the root cause. You need a hash function (mapping to Int64 for performance) with the property `h(v)+h(w) == h(v+w)`. Furthermore, no collisions are allowed. 

If these are the requirements, then I would have the following proposals:

- Can we check in `set_variables` for collisions by `length.(pos_table) == length.(index_table)`? This should take no time and would ensure that we don't have hash collisions and therefore no wrong calculations.
- I think the modulo arithmetic is in general not a problem, as long as we don't have collisions. I originally run into the issue with `order=1` and `numvars=70`. For `order=1` you use a hash function in base `2` that does not nicely work with modulo 2^64. What if we always choose an even base, then I would imagine that collisions are less likely. With this small modification I don't find any collision for `order=1:10` and `numvars=1:100` if the series has less than 20k terms (to keep it reasonable ;) ), see code below.

```
using TaylorSeries
import TaylorSeries: in_base, pos_table, index_table, coeff_table

function TaylorSeries.in_base(order, v)
    order = order+1
    if iseven(order)  # NEW
        order += 1
    end

    result = 0

    for i in v
        result = result*order + i
    end

    result
end

function taylor_element_count(o, n)
    count = 0
    for i=0:o
        count += binomial(n + i - 1, i)
    end
    return count
end 

function hascollision(o, n)
    set_variables("x", order=o, numvars=n)
    return length.(pos_table) != length.(index_table)
end

for o=1:10, n=1:100
    if taylor_element_count(o, n) < 20000
        if hascollision(o, n)
            @show o, n
        end
    end
end

```

**Comment by lbenet** [comment]
Can you provide a bit more details of the problem you have; my guess is that your case (`numvars=70`, `order=1`) should work...

**Comment by lbenet** [comment]
Sorry, you are right, I see it now, there is indeed a problem there:
```julia
julia> any(keys(TS.pos_table[2]) .<= 0)
true

julia> TS.pos_table[2][-9223372036854775808]
7
```

**Comment by lbenet** [comment]
I just opened #314; while it doesn't solve the problem, it avoids being silent with inconsistent tables.

**Comment by lbenet** [comment]
@patkwee If this solution is good enough, at least for the time being, I'll go ahead and merge...

**Comment by patkwee** [comment]
It is definitely an improvement to throw an error instead of having a silent error.

Though, since I would like to use more than 65 variables with order one, it is, as you stated, not solving my problem.

Can you explain what the problem with negative or overflown index_table entries is in case there are no hash collisions? I looked a bit into where and how index_table is used, and don’t recognize an issue with negative hash values or modulo arithmetic.

**Comment by lbenet** [comment]
A `TaylorN{T}` ((of specified `order`) is essentially a `Vector{HomogeneousPolynomial{T}}`, with the indices of the vector identifying the degree of the `HomogeneousPolynomial`: `index = degree+1`, and `degree` runs from 0 to `order`. Each `HomogeneousPolynomial{T}` is itself a `Vector{T<:Number}`, now the indices identifying the coefficients in front of the monomials of degree of the `HomogeneousPolynomial`. 

To keep track of the monomials for all possible degrees, we use 4 different tables (see [TaylorSeries.generate_tables](https://juliadiff.org/TaylorSeries.jl/stable/api/#TaylorSeries.generate_tables)), which are `Vector{something}`. The `index_table` has, in its `i+1` entry, the hashes of the *ordered* list of monomials of degree `i`; `pos_table[i+1]` links those hashes (through a `Dict`) with the actual position (the indices of an array) of the monomials that build the `Homogeneouspolynomial`. The `coeff_table` is a `Vector{Vector{Int}}` that ontains in its `i+1` entry the exponents in each variable that build each monomial of degree `i`, in a specific order. (`pos_table` and `index_table` are essentially the inverse of the other.) The hashes correspond to an *integer value*, in base `order+1`, with `numvar` entries ("digits"), where each entry represents the exponent with respect to a variable, all of them building the exponents of the monomial. (The fourth table, `size_table` is actually irrelevant.)

For example, you are multiplying two monomials, resulting in $x^3 y^1 z^2$ and you have set the parameters `order=6` and `numvars=3`. Then, that monomial is hashed as "321_base7", i.e. `3*7^2+1*7+2 = 156`, which you obtain as `TaylorSeries.in_base(6,[3,1,2])`. It's position in the arrangement of the coefficients of the `HomogeneousPolynomial` is the `TaylorSeries.pos_table[7][156] = 9`-th entry. Note that that monomial can be obtained as the product of $x z$ and $x^2 y z$. If you locate those monomials, you get the hashes `50` and `106`, which add up to `156`, just as the exponents add to produce the monomial $x^3 y^1 z^2$.

Now, your question: if you get a negative number (due to the modal arithmetic), or if 0 appears twice as a hash, there is no way to ensure that the last property holds, namely, the hash associated to the product of two monomials may not correspond to adding the hashes of the monomials. This is the actual problem.

**Comment by lbenet** [comment]
@patkwee #314 is merged, a new version released. I'm keeping this issue open, to recall it. If I have some time, I'll try to at least make it work with larger parameter values...

**Comment by patkwee** [comment]
@lbenet Thanks for this elaborate explanation, I really appreciate this and it confirms my previous understanding how the hashes and tables are used. 

Though, I think your statement is not correct:
> there is no way to ensure that the last property holds, namely, the hash associated to the product of two monomials may not correspond to adding the hashes of the monomials. This is the actual problem.

Let me explain a bit. You require a hash function with `h(u + v) = h(u) + h(v)` (exactly as you explained) and which has no collisions, i.e. if `h(u) == h(v)` then `u == v`. I fully recognize the issue with collisions in the current implementation, but let's ignore this for the moment, I will come back to this.

I believe the modular arithmetic with Int64 is not breaking the `h(u + v) = h(u) + h(v)` property. I am not expert on this, but reading https://en.wikipedia.org/wiki/Modular_arithmetic , we are essentially using the *compatibility with addition* property.

Consider the following example for `order=2`, `numvars=42`:

```julia
julia> BigInt(3)^40 > typemax(Int)
true

julia> (BigInt(3)^41 + BigInt(3)^40) % 2^63
2513801651953836164

julia> ((BigInt(3)^41 % 2^63) + (BigInt(3)^40 % 2^63)) % 2^63
2513801651953836164
```

Here is another example using `in_base` function:

```julia
using TaylorSeries
import TaylorSeries: in_base

# Example in order=2, numvars=42

u = zeros(Int64, 42)
u[20] = 1

v = zeros(Int64, 42)
v[1] = 1

w = zeros(Int64, 42)
w[2] = 1


@show in_base(2, u)  # no overflow, 3^(42-20) < typemax(Int64)
@show in_base(2, v)  # overflow, 3^41 > typemax(Int64)
@show in_base(2, w)  # overflow, 3^40 > typemax(Int64)

# checking required hash property h(a+b) = h(a) + h(b)
check(a, b) = in_base(2, a + b) == in_base(2, a) + in_base(2, b), in_base(2, a + b)

@show check(u, u)  # true
@show check(u, v)  # true
@show check(u, w)  # true
@show check(v, u)  # true
@show check(v, v)  # true
@show check(w, w)  # true
```

So I think, an overflow and negative hash values are not a problem for the `h(u + v) = h(u) + h(v)` property. If you have a counter example, this would be very helpful.

There is obviously an issue if we have collisions, for example:

```julia
# Example in order=1, numvars=66

u = zeros(Int64, 66)
u[1] = 1

v = zeros(Int64, 66)
v[2] = 1

in_base(1, u)  # =0
in_base(1, v)  # =0
```

Here `in_base(1, u) == in_base(1, v)` but `u != v`.

If we avoid overflows in `in_base` we are also sure that there is no collision, as you proposed in #314 (BTW you could have also simply checked for `log(order+1, typemax(Int64)) > numvars`) . At the same time I think this is an unnecessary restrictive check. 

My proposal to check for `length.(pos_table) == length.(index_table)` also ensures that there are no collisions. If the number of entries in `pos_table` is less than in `index_table`, one or more monomials where mapped to the same hash value.

If we further avoid an even base in `in_base`, by just increasing it when it is even, we can avoid many collisions.

```julia
function TaylorSeries.in_base(order, v)
    order = order+1
    if iseven(order)  # NEW
        order += 1
    end

    result = 0

    for i in v
        result = result*order + i
    end

    result
end
```

You could argue that this will make the hash value unnecessary large, though, since we should not care about overflow and should only care about collisions, this looks like the better choice.

To further illustrate this, in my application I am mostly interested in `order=1` and `order=2` for many variables. With this simple tweak, I was not able to find a collision:

```julia
## Check for order=1 collisions
s = Set{Int64}()
push!(s, 0)  # hash of constant term
N = Int(10e6)

for i=1:N
    h = 3^i
    if h in s
        @show "Collision", i, h
        break
    end
    push!(s, h)
end


## Check for order=2 collisions
s = Set{Int64}()
push!(s, 0)  # hash of constant term
N = Int(1e4)

# first order terms
for i=1:N
    h = 3^i
    if h in s
        @show "Collision", i, h
        break
    end
    push!(s, h)
end

# second order terms
for i=1:N, j=i:N
    h = 3^i + 3^j
    if h in s
        @show "Collision", i, j, h
        break
    end
    push!(s, h)
end
```

I searched up to 1e7 variables in `order=1` and up to 1e4 variables in `order=2` (BTW this is more than enough for my application).

I hope I could explain a bit better my proposed solution https://github.com/JuliaDiff/TaylorSeries.jl/issues/85#issuecomment-1434583842. Let me know what you think.

**Comment by lbenet** [comment]
I agree *partially* with you 😄. 

If you can be sure the uniqueness of the labels holds, there should be no problem. Yet, modal arithmetic is the responsible for `julia> 2^64 == 2^65 == 2^66 == 0`, which breaks the uniqueness of the labels. Another case occurs for order 3 and 33 variables (4^32 == 4^33 == 0). I do not know if this happens only for odd values of `order`, but it is cumbersome to check.

Regarding your proposal, I think `length.(pos_table) == length.(index_table)` does not truly checks for collisions (uniqueness of the labels). I think `allunique.(index_table)` does:
```julia
julia> set_variables("x", numvars=70, order=1);

julia> all(allunique.(TS.index_table))
false

julia> set_variables("x", numvars=200, order=2);

julia> all(allunique.(TS.index_table))
true

julia> set_variables("x", numvars=35, order=3);

julia> all(allunique.(TS.index_table))
false
```

**Comment by patkwee** [comment]
My proposal consists of two aspects: 1) to introduce a check for uniqueness (to avoid silent errors) and 2) to improve *uniqueness* by using only odd bases in `in_base`.

1) I like your suggestion to use `all(allunique.(index_table))` to check for uniqueness. It is more expressive than my original proposal.  (I still think `length.(pos_table) == length.(index_table)` would also do the job.)

2) By adding `if iseven(order)  order += 1  end` in `in_base` I think we can drastically improve uniqueness (i.e. avoiding hash collisions). As I showed in the previous post, for `order=1`, where we would then use base of 3 instead of 2, even for 1e7 variables all labels would be unique (compared to base 2 where it will break down for >65 variables). 

A more conservative approach might be to only change the base for `order=1`, where I showed that it will improve uniqueness and will solve the issue for my application, i.e. something like:

```julia
function TaylorSeries.in_base(order, v)
    if order == 1
       base = 3   # NEW
    else 
       base = order+1
    end

    result = 0

    for i in v
        result = result*base + i
    end

    result
end
```

Though, then hash collisions for, for example, `numvars=32`, `order=3` would still exist and will throw an error. Therefore, I would still favor to avoid even bases completely. I don't see a drawback of this change.
